3rdparty/openvx/include/ivx.hpp

```diff
@@ -0,0 +1,1611 @@
+    {
+        if (isMapped()) throw WrapperError(std::string(__func__)+"(): already mapped");
+        if(!_externalMem) free();
+        _data = 0; // could still point to external
+#ifdef VX_VERSION_1_1
+        IVX_CHECK_STATUS(vxMapImagePatch(img, &rect, planeIdx, &_mapId, &_addr, &_data, usage, _memType, flags) );
+#else
+        IVX_CHECK_STATUS(vxAccessImagePatch(img, &rect, planeIdx, &_addr, &_data, usage));
+        _rect = rect;
+        _planeIdx = planeIdx;
+#endif
+        _img = img;
+        _externalMem = true;
+#ifdef IVX_USE_OPENCV
+        _m.release();
+#endif
+    }
+
+    void unmap()
+    {
+#ifdef VX_VERSION_1_1
+        IVX_CHECK_STATUS(vxUnmapImagePatch(_img, _mapId));
+#else
+        IVX_CHECK_STATUS(vxCommitImagePatch(_img, &_rect, _planeIdx, &_addr, _data));
+        _rect.start_x = _rect.end_x = _rect.start_y = _rect.end_y = 0u;
+        _planeIdx = -1;
+
+#endif
+        _img = 0;
+        _data = 0;
+        _externalMem = false;
+#ifdef IVX_USE_OPENCV
+        _m.release();
+#endif
+    }
+
+    ~Patch()
+    {
+        try { if (_img) unmap(); } catch(...) {; /*ignore*/}
+        if(!_externalMem) free();
+    }
+
+    void* pixelPtr(vx_uint32 x, vx_uint32 y)
+    {
+        if (!_data) throw WrapperError(std::string(__func__)+"(): base pointer is NULL");
+        if (x >= _addr.dim_x) throw WrapperError(std::string(__func__)+"(): X out of range");
+        if (y >= _addr.dim_y) throw WrapperError(std::string(__func__)+"(): Y out of range");
+        return vxFormatImagePatchAddress2d(_data, x, y, &_addr);
+    }
+
+private:
+    Patch(const Patch& p); // = delete
+    Patch& operator=(const Patch&); // = delete
+#ifndef IVX_USE_CXX98
+    Patch& operator=(Patch&&); // = delete
+#endif
+};
+
+/*
+* Param
+*/
+class Param : public RefWrapper<vx_parameter>
+{
+public:
+    IVX_REF_STD_CTORS_AND_ASSIGNMENT(Param);
+    // NYI
+};
+
+/*
+* Scalar
+*/
+class Scalar : public RefWrapper<vx_scalar>
+{
+public:
+    IVX_REF_STD_CTORS_AND_ASSIGNMENT(Scalar);
+
+    static Scalar create(vx_context c, vx_enum dataType, const void *ptr)
+    { return Scalar( vxCreateScalar(c, dataType, ptr) ); }
+
+    template<typename T> static Scalar create(vx_context c, vx_enum dataType, T value)
+    { return Scalar( vxCreateScalar(c, dataType, &value) ); }
+
+    template<vx_enum E> static Scalar create(vx_context c, typename EnumToType<E>::type value)
+    { return Scalar( vxCreateScalar(c, E, &value) ); }
+
+#ifndef VX_VERSION_1_1
+static const vx_enum VX_SCALAR_TYPE = VX_SCALAR_ATTRIBUTE_TYPE;
+#endif
+    vx_enum type()
+    {
+        vx_enum val;
+        IVX_CHECK_STATUS( vxQueryScalar(ref, VX_SCALAR_TYPE, &val, sizeof(val)) );
+        return val;
+    }
+
+    template<typename T>
+    void getValue(T& val)
+    {
+        vx_enum t = type();
+        if(sizeof(T) != enumToTypeSize(t)) throw WrapperError(std::string(__func__)+"(): incompatible types");
```

Comment:
Don't you think that `TypeToEnum<T>::value == type()` requirement is too strong?